Remove notion of Modification

ifesdjeen opened the pull request:
Patch for CASSANDRA-19117 by Alex Petrov; reviewed by Abe Ratnofsky

ifesdjeen committed:
Remove notion of Modification

Patch for CASSANDRA-19117 by Alex Petrov; reviewed by Abe Ratnofsky

maedhroz commented on harry-core/src/harry/core/Configuration.java:
```diff
@@ -885,20 +885,20 @@ public DefaultCDSelectorConfiguration build()
         @JsonCreator
         public DefaultCDSelectorConfiguration(@JsonProperty("modifications_per_lts") DistributionConfig modifications_per_lts,
-                                              @JsonProperty("rows_per_modification") DistributionConfig rows_per_modification,
+                                              @JsonProperty("rows_per_modification") DistributionConfig operations_per_lts,
```
Leaving "rows_per_modification" for back-compat?

ifesdjeen replied:
Removed rows_per_modification now and cleaned up modifications_per_lts. Missed this.

maedhroz commented on harry-core/src/harry/core/Configuration.java:
```diff
@@ -885,20 +885,20 @@ public DefaultCDSelectorConfiguration build()
     public static class DefaultCDSelectorConfiguration implements CDSelectorConfiguration
     {
         public final DistributionConfig modifications_per_lts;
```
Is `modifications_per_lts` here unused now?

ifesdjeen replied:
Yes; removed

maedhroz commented on harry-core/src/harry/dsl/HistoryBuilder.java:
```diff
@@ -361,32 +350,22 @@ public HistoryBuilder finish()
             Counter opId = new Counter();
             for (Step step : steps)
             {
-                batches.add(step.toBatch(pd, lts, m.get(), opId::getAndIncrement));
```
Cool, so we no longer need the notion of a modification ID?

ifesdjeen replied:
yes! Is a small simplificaiton, but makes things much simpler when working on history builder, less indirection.

maedhroz commented on harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java:
```diff
@@ -72,44 +78,43 @@ public InJVMTokenAwareVisitExecutor(Run run,
+                Uninterruptibles.sleepUninterruptibly(delaySecs, TimeUnit.SECONDS);
+            }
         }
+        throw new IllegalStateException(String.format("Can not execute statement %s after %d retries", statement, retries));
```
So tl;dr we made this synchronous and added some logging for visibility? I'll figure out why I suppose as I look at its callers...

ifesdjeen replied:
Let's say this was the simplest way to know that a query _has_ in fact finished. This is a barebone model, and probably should be improved upon. But at least it gives you some ground to stand upon.

maedhroz commented on harry-core/src/harry/model/DescriptorSelectorBuilder.java:
```diff
@@ -33,7 +33,7 @@ public class DescriptorSelectorBuilder implements Configuration.CDSelectorConfig
     private OpSelectors.OperationSelector operationSelector = DEFAULT_OP_SELECTOR;
-    private Distribution numberOfRowsDistribution = new Distribution.ScaledDistribution(2, 30);
+    private Distribution modificationsPerLts = new Distribution.ScaledDistribution(2, 30);
     private Distribution numberOfModificationsDistribution = new Distribution.ScaledDistribution(1, 3);
```
`numberOfModificationsDistribution` is unused now?

ifesdjeen replied:
Good point. Removed!

maedhroz commented on harry-core/src/harry/model/DescriptorSelectorBuilder.java:
```diff
@@ -72,9 +72,9 @@ public DescriptorSelectorBuilder setOperationSelector(OpSelectors.OperationSelec
      * we'll always be limited by this.
      **/
-    public DescriptorSelectorBuilder setRowsPerModificationDistribution(Distribution numberOfRowsDistribution)
+    public DescriptorSelectorBuilder setModificationsPerLts(Distribution numberOfRowsDistribution)
```
```suggestion
    public DescriptorSelectorBuilder setModificationsPerLts(Distribution this.modificationsPerLts)
```
?

maedhroz replied:
I guess we've named it `operationsPerLtsDistribution` elsewhere, so I guess that's fine too if we want to be explicit.

ifesdjeen replied:
Renamed to `setOperationsPerLts`

maedhroz commented on harry-core/src/harry/model/DescriptorSelectorBuilder.java:
```diff
@@ -72,9 +72,9 @@ public DescriptorSelectorBuilder setOperationSelector(OpSelectors.OperationSelec
      * In a given modification, we are only able to visit as many rows as there are rows in the partition, so
      * we'll always be limited by this.
```
nit: Comment needs tweaking?

maedhroz commented on harry-core/src/harry/model/DescriptorSelectorBuilder.java:
```diff
@@ -33,7 +33,7 @@ public class DescriptorSelectorBuilder implements Configuration.CDSelectorConfig
+    private Distribution modificationsPerLts = new Distribution.ScaledDistribution(2, 30);
     private Distribution numberOfModificationsDistribution = new Distribution.ScaledDistribution(1, 3);
     private int maxPartitionSize = Integer.MAX_VALUE;
     private Function<SchemaSpec, int[]> fractionsSupplier = null;
```
nit: What does `fractionsSupplier` do?

ifesdjeen replied:
The initial idea was to use HierarchicalDescriptorSelector: sort of to make "predictable" partitions, where parts of CK would repeat.

maedhroz commented on harry-core/test/harry/model/OpSelectorsTest.java:
```diff
@@ -73,19 +72,16 @@ public void testRowDataDescriptorSupplier()
             long pd = pdSupplier.pd(lts);
-            for (int m = 0; m < descriptorSelector.numberOfModifications(lts); m++)
+
+            int opsPerMod = descriptorSelector.operationsPerLts(lts);
```
```suggestion
            int opsPerLts = descriptorSelector.operationsPerLts(lts);
```

ifesdjeen replied:
Great point!

maedhroz commented on harry-core/test/harry/operations/RelationTest.java:
```diff
@@ -183,12 +183,12 @@ public BitSet columnMask(long pd, long lts, long opId, OpSelectors.OperationKind
                                                                       }
 
-                                                                      public long modificationId(long pd, long cd, long lts, long vd, int col)
+                                                                      public long operationId(long pd, long cd, long lts, long vd, int col)
```
unused now?

ifesdjeen replied:
Good point! Removed this and some more.

maedhroz reviewed (commented):
Mostly looks good. Just dropped some minor comments/cleanup requests.

Is `CDSelectorConfigurationBuilder` going to retain `modifications_per_lts` and `rows_per_modification`?

ifesdjeen commented:
Removed modifications_per_lts and rows_per_modification and introduced operations_per_lts.

I _think_ most of your comments should now be addressed.

aratno commented on harry-core/src/harry/concurrent/Uninterruptibles.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.concurrent.TimeUnit;
+
+public class Uninterruptibles
```
Avoiding using Guava?
